### PR TITLE
[FEAT] User 탈퇴 기능 구현 

### PR DESCRIPTION
**Pull request by 5hseok**
## #️⃣ Related Issue

Linked Feature backlog: 

## 📝 Purpose of PR

Please briefly explain your work in this PR.

## Contents

List of work items performed.

## Checklist

-   [ ] Does commit messages follow the convention?
-   [ ] Are variable names brief but descriptive?
-   [ ] Is the code clean and easy to understand?
-   [ ] If a new feature has been added, or a bug fixed, has a test been added to confirm good behavior?
-   [ ] Does the test successfully test edge and corner cases?

## 💬 Review Requirements (Optional)

Is there anything you can praise about this PR? Start with praise.
Make a review, leaving kind comments and suggesting changes where needed (to resolve the above).

### CODE REVIEW

File: src/test/java/ac/dnd/dodal/acceptance/auth/withdrawUserAcceptanceTest.java
Comment (5hseok) [inline src/test/java/ac/dnd/dodal/acceptance/auth/withdrawUserAcceptanceTest.java:1]: 탈퇴 처리한 사용자가 다시 요청을 보냈을 때 Reject 당하는 것도 테스트 코드에 추가하려고 했는데, 다른 클래스의 테스트들에도 영향이 가는 문제를 해결하지 못해서 탈퇴 테스트만 작성했습니다.

실제로 요청을 보내 테스트해보니 탈퇴 처리 이후에는 다시 접근도 안되고 다시 재로그인했을 때도 원하는 동작이 되는 것을 확인했습니다.

File: src/main/java/ac/dnd/dodal/application/user/service/AuthLoginService.java [hunk 1]
```diff
@@ -107,13 +104,32 @@ private UserInfoResponseDto buildUserInfoResponse(User user, JwtTokenDto jwtToke
   }
 
   private User registerNewAppleUser(OAuthByAppleUserInfoRequestDto appleIdTokenParsingDto) {
-    User newUser = new User(appleIdTokenParsingDto.email(), null, appleIdTokenParsingDto.deviceToken(), appleIdTokenParsingDto.appleId(), UserRole.USER);
-    return userRepository.save(newUser);
+    // 탈퇴한 이력이 있는 사용자인지 확인
+    User withdrawnUser = checkWithdrawnUser(appleIdTokenParsingDto.appleId());
+
+    if (withdrawnUser != null && withdrawnUser.getDeletedAt() != null) {
```
Comment (5hseok) [inline src/main/java/ac/dnd/dodal/application/user/service/AuthLoginService.java:110]: 탈퇴한 기록이 있는 사용자가 있다면 해당 사용자의 deletedAt을 null로 초기화합니다.

File: src/main/java/ac/dnd/dodal/application/user/service/AuthLoginService.java [hunk 2]
```diff
@@ -107,13 +104,32 @@ private UserInfoResponseDto buildUserInfoResponse(User user, JwtTokenDto jwtToke
...
+      withdrawnUser.reactivateDeletedUser();
+      return withdrawnUser;
+    }
+    else {
```
Comment (sueelly) [inline src/main/java/ac/dnd/dodal/application/user/service/AuthLoginService.java:114]: 위의 if문에서 return을 수행하니 else는 없어도 될것같습니다

File: src/main/java/ac/dnd/dodal/domain/user/model/User.java [hunk 3]
```diff
@@ -65,7 +69,16 @@ public void updateUserProfileImageUrl(String profileImageUrl) {
...
+    public void withdrawUser(){
+        this.deletedAt = LocalDateTime.now();
+        this.refreshToken = null;
+    }
+
     public void deleteUserProfileImageUrl() {
         this.profileImageUrl = null;
     }
+
+    public void reactivateDeletedUser(){
```
Comment (sueelly) [inline src/main/java/ac/dnd/dodal/domain/user/model/User.java:81]: 탈퇴한 유저에 대해서 withdrawn user, deleted user 2가지 단어가 혼용되어 사용되는 것 같은데 하나로 정해서 통일하는 게 좋을 것 같습니다~

File: src/main/java/ac/dnd/dodal/application/user/service/AuthLoginService.java [hunk 4]
```diff
@@ -107,13 +104,32 @@ private UserInfoResponseDto buildUserInfoResponse(User user, JwtTokenDto jwtToke
...
+    }
   }
 
   private User registerNewKakaoUser(KakaoUserInfoDto kakaoUserInfo, String deviceToken) {
     User newUser =
         new User(kakaoUserInfo.nickname(), kakaoUserInfo.profileImageUrl(), deviceToken, kakaoUserInfo.email(), UserRole.USER);
     return userRepository.save(newUser);
   }
+
+  private User checkWithdrawnUser(String email) {
```
Comment (sueelly) [inline src/main/java/ac/dnd/dodal/application/user/service/AuthLoginService.java:132]: 탈퇴한 유저인지 확인하는 로직이 이 메서드에 들어가야할 것 같아요~
boolean을 반환하고, isWithdrawnUser로 수정하는 형식은 어떨까요?

File: src/main/java/ac/dnd/dodal/application/user/service/UserCommandService.java [hunk 5]
```diff
@@ -116,4 +122,20 @@ public void createUserAnswer(Long userId, CreateUserAnswerRequestDto createUserA
 
         eventPublisher.publishEvent(new OnboardingProceededEvent(user.getId(), userAnswers));
     }
+
+    @Override
+    public DeleteUserInfoResponseDto withdrawUser(Long userId){
+
```
Comment (sueelly) [inline src/main/java/ac/dnd/dodal/application/user/service/UserCommandService.java:128]: nit: 불필요한 공백

File: src/main/java/ac/dnd/dodal/application/user/service/UserCommandService.java [hunk 6]
```diff
@@ -116,4 +122,20 @@ public void createUserAnswer(Long userId, CreateUserAnswerRequestDto createUserA
...
+        try {
+        User user = userCommandRepository.findById(userId)
+                .orElseThrow(() -> new UserNotFoundException(UserExceptionCode.NOT_FOUND_USER));
+
+        user.withdrawUser();
+
+        return DeleteUserInfoResponseDto.fromUserEntity(user);
+
+        } catch (Exception e){
```
Comment (sueelly) [inline src/main/java/ac/dnd/dodal/application/user/service/UserCommandService.java:137]: 여기서 Exception으로 catch하면
UserNotFoundException이 `InternalServerErrorException(UserExceptionCode.DELETE_FAILED)`으로 바뀌어서 다시 throw되는 상황입니다

UserNotFound는 그 예외 그대로 전달되어야한다고 생각하는데 변환을 해야하는 이유가 있을지,
다른 예외를 catch하여 throw하고 싶으신거였다면 어떤 예외가 발생할 수 있다고 생각하신건지 궁금합니다

File: src/main/java/ac/dnd/dodal/ui/auth/response/DeleteUserInfoResponseDto.java
Comment (sueelly) [inline src/main/java/ac/dnd/dodal/ui/auth/response/DeleteUserInfoResponseDto.java:1]: 회원 탈퇴 시에 해당 데이터를 전달하는 이유가 무엇인가요?
보통 성공 / 실패 정도만 전달해서요

File: src/main/java/ac/dnd/dodal/ui/auth/response/UserInfoResponseDto.java [hunk 7]
```diff
@@ -6,13 +6,12 @@
...
 @Builder(access = AccessLevel.PRIVATE)
 public record UserInfoResponseDto(
-    String nickname, String email, String profileImageUrl, boolean checkOnboardingSuccess, JwtTokenDto jwtTokenDto) {
+    String nickname, String email, String profileImageUrl, JwtTokenDto jwtTokenDto) {
   public static UserInfoResponseDto fromUserEntity(User user, JwtTokenDto jwtTokenDto) {
     return UserInfoResponseDto.builder()
         .nickname(user.getNickname())
         .email(user.getEmail())
         .profileImageUrl(user.getProfileImageUrl())
-            .checkOnboardingSuccess(true)
```
Comment (sueelly) [inline src/main/java/ac/dnd/dodal/ui/auth/response/UserInfoResponseDto.java:15]: 변경사항 보았는데, 온보딩이 완료되어야하는 것은 동일하지 않나요?

만약 온보딩이 완료되지 않아도 동작하도록 만들어야하면 여기서 true를 넘기는 게 아니라 해당 로직을 체크하는 부분에서 값을 다르게 주는 게 적절할 것 같습니당.

File: src/main/java/ac/dnd/dodal/domain/user/model/User.java [hunk 8]
```diff
@@ -55,6 +55,10 @@ public void updateRefreshToken(String refreshToken) {
         this.refreshToken = refreshToken;
     }
 
+    public void updateUserRole(UserRole role) {
```
[the remark this answers is not in the document]
Comment (sueelly) [reply src/main/java/ac/dnd/dodal/domain/user/model/User.java:58]: 좋습니다~

그렇다면 현재 이 메서드가 사용되는 곳이 있나요?
사용될 일이 없을 것 같아서요

**Review by sueelly — CHANGES_REQUESTED** [review]
이제 급한 불은 꺼졌으니 다시 PR해야만 merge하도록할까요?

여기는 pr로 올라왔는데 fork한 것은 이미 반영이 되어있어서 변경사항 생기면 제가 올릴 때도 충돌날것같네용🥲

File: src/main/java/ac/dnd/dodal/domain/user/model/User.java [hunk 9]
```diff
@@ -65,7 +69,16 @@ public void updateUserProfileImageUrl(String profileImageUrl) {
         this.profileImageUrl = profileImageUrl;
     }
 
+    public void withdrawUser(){
+        this.deletedAt = LocalDateTime.now();
+        this.refreshToken = null;
+    }
```
[the remark this answers is not in the document]
Comment (sueelly) [reply src/main/java/ac/dnd/dodal/domain/user/model/User.java:75]: 그 안에서 delete 메서드를 호출하는 것도 좋을 것 같아요~ 둘이 동일한 로직이라서요

File: src/test/java/ac/dnd/dodal/acceptance/auth/withdrawUserAcceptanceTest.java
Comment (sueelly) [reply src/test/java/ac/dnd/dodal/acceptance/auth/withdrawUserAcceptanceTest.java:1]: 추가된 user 샘플 데이터 중에 인수 테스트에서 쓰이지 않는 것을 찾아서 쓰셔도 될것같아용

File: src/main/java/ac/dnd/dodal/ui/auth/AuthController.java [hunk 10]
```diff
@@ -42,4 +48,10 @@ public ApiResponse<UserInfoResponseDto> refresh(
       @NotNull @RequestHeader(Constants.AUTHORIZATION_HEADER) String refreshToken) {
     return ApiResponse.success(authService.refresh(refreshToken));
   }
+
+  @DeleteMapping("/withdraw")
+    public ApiResponse<DeleteUserInfoResponseDto> softDeleteUser(
```
Comment (sueelly) [inline src/main/java/ac/dnd/dodal/ui/auth/AuthController.java:53]: 여기서 유저 정보를 반환하는 이유가 있나요?
왜 필요한지 설명해주시면 감사하겠습니다~

그리고 harddelte하는 로직을 따로 둘 것이 아니면 메서드명은 deleteUser로 두어도 충분할 것 같습니다
+ DeletedUserResponse로 pp 형태가 되어야할 것 같습니당

**Review by sueelly — CHANGES_REQUESTED** [review]
이해가 잘 안되는 부분들이 있어서 설명 해주시면 감사하겠습니다~

**Review by 5hseok — COMMENTED** [review]
근데 생각해보니까 지금 방식으로는 사용자가 탈퇴했다가 다시 가입하면 기존 데이터를 그대로 이어서 사용하는 게 되는데, 저희 정책이 탈퇴하면 기존 데이터를 이어서 사용하는 것과 아예 새로 시작하는 것 중에 어떤 것이어야 할까요?

전자라면 괜찮을 것 같은데, 후자라면 User와 관련된 테이블 레코드들을 모두 delete() 해줘야 할 것 같아서요

**Comment by sueelly** [comment]
원래 탈퇴 정책 같은 걸 따로 만드는데, 보편적으로 모든 기록을 날리는 게 맞는 것 같아요!
user delete event 생성해서 비동기로 관련 데이터들 삭제하고,
후에 스케쥴러 생성해서 제대로 삭제되었는지 확인 & 남은 것 삭제하는 것 추가한다음
+ 일정주기마다 soft delete -> hard delete하는 것까지 구현해야할것같네용

**Review by sueelly — APPROVED** [review]

**Comment by sueelly** [comment]
그리고 해당 JPA 예외와 같은 것은 어떤 도메인에서 에러가 났는지 굳이 표시할 필요가 없을 것 같아요.
어차피 정의한 예외들이 아니면 모두 Internal Server Error로 반환될거고,
예외 메세지는 결국 client에게 전달하기 위함인데 이런 경우 도메인이 중요한 건 아닐 것 같아서요.
